File: minion/app/src/main/java/snakesoft/minion/Models/GlobalModel.java
package snakesoft.minion.Models;

import android.content.Context;
import android.os.AsyncTask;
import android.widget.Toast;

import java.io.UnsupportedEncodingException;
import java.util.Vector;

import snakesoft.minion.Materia.InboxServiceProxy;
import snakesoft.minion.Materia.CalendarServiceProxy;
import snakesoft.minion.Materia.MateriaConnection;
import snakesoft.minion.Materia.MateriaUnreachableException;
import snakesoft.minion.Activities.SyncListener;

/**
 * Created by snake on 4/28/17.
 */

public class GlobalModel
{
    public static SyncObserver getSyncObserver()
    {
        return mSyncObserver;
    }

    static class SyncAllTask extends AsyncTask<SyncListener, Void, SyncListener>
    {
        protected SyncListener doInBackground(SyncListener... p)
        {
            mSyncResult = GlobalModel.doSync();
            return p[0];
        }

        protected void onPostExecute(SyncListener listener)
        {
            if(mSyncResult)
            {
                listener.onSyncComplete();
            }
            else
            {
                listener.onSyncError();
            }
        }

        private boolean mSyncResult;
    }

    protected static boolean doSync()
    {
        try
        {
            mSyncObserver = new SyncObserver();

            mInboxModel.sync(mSyncObserver);
            mCalendarModel.sync(mSyncObserver);

            return true;
        }
        catch(MateriaUnreachableException ex)
        {
            return false;
        }
    }

    public static void init(Context context)
    {
        mLocalDatabase = new LocalDatabase(context);

        byte[] ipbytes = mLocalDatabase.get("Settings.IP");
        if(ipbytes != null)
        {
            String ip = null;
            try
            {
                ip = new String(ipbytes, "UTF-8");
            }
            catch (UnsupportedEncodingException e)
            {

            }
            mConnection = new MateriaConnection(ip);
        }
        else
        {
            mConnection = new MateriaConnection("localhost");
        }

        mInboxModel = new InboxModel(new InboxServiceProxy(mConnection));
        mWpModel = new WpModel();
        mCalendarModel = new CalendarModel(
                new CalendarServiceProxy(mConnection));

        loadState(context);
    }

    private static void loadState(Context context)
    {
        boolean dbok = true;
        try
        {
            mInboxModel.loadState(mLocalDatabase);
        } catch (Exception e)
        {
            dbok = false;
        }
        try
        {
            mCalendarModel.loadState(mLocalDatabase);
        } catch (Exception e)
        {
            dbok = false;
        }

        mWpModel.loadState(mLocalDatabase);

        if(!dbok)
        {

            CharSequence text = "Database is inconsistent please reset.";
            int duration = Toast.LENGTH_LONG;

            Toast toast = Toast.makeText(context, text, duration);
            toast.show();
            return;
        }
    }

    public static void sync(SyncListener listener)
    {
        new SyncAllTask().execute(listener);
    }

    public static InboxModel getInboxModel()
    {
        return mInboxModel;
    }
    public static CalendarModel getCalendarModel() {return mCalendarModel; }
    public static WpModel getWpModel() {return mWpModel;}

    public static String getIp()
    {
        return mConnection.getIp();
    }

    public static void reset()
    {
        mInboxModel.resetChanges();
        mCalendarModel.resetChanges();
    }

    public static void setNewIp(String ip)
    {
        if(!mConnection.getIp().equals(ip))
        {
            mLocalDatabase.put("Settings.IP", ip.getBytes());
            mConnection.setNewIp(ip);
            reset();
        }
    }

    static private MateriaConnection mConnection;
    static private LocalDatabase mLocalDatabase;
    static private InboxModel mInboxModel;
    static private CalendarModel mCalendarModel;
    static private WpModel mWpModel;
    static private SyncObserver mSyncObserver;
}

File: minion/app/src/main/java/snakesoft/minion/Models/GlobalModel.kt
package snakesoft.minion.Models

import snakesoft.minion.materia.CalendarServiceProxy
import snakesoft.minion.materia.JournalServiceProxy
import snakesoft.minion.materia.MateriaConnection
import snakesoft.minion.materia.MateriaUnreachableException

object GlobalModel
{
    var Ip: String = "62.171.175.23"
        set(value)
        {
            field = value
            Db.put("Settings.IP", value)
            reset()
        }

    private val Db = LocalDatabase()
    val InboxModel = InboxModel(Db)
    val CalendarModel = CalendarModel(Db)
    val JournalModel = JournalModel(Db)
    val FinanceModel = FinanceModel(Db)
    val RewardModel = RewardModel(Db)
    var NewsModel = NewsModel(Db)

    internal fun doSync(syncObserver: SyncObserver, password: String): Boolean
    {
        return try
        {
            val connection = MateriaConnection(Ip, password, "5757")
            val connectionNew = MateriaConnection(Ip, password, "5756")

            //check password
            val p = JournalServiceProxy(connectionNew)
            if(p.loadIndex().count() == 0)
            {
                syncObserver.OnUpdated("Incorrect password")
                false
            }
            else
            {
                RewardModel.sync(syncObserver, connection)


                InboxModel.sync(syncObserver, connectionNew)
                CalendarModel.sync(syncObserver, connectionNew)
                FinanceModel.sync(syncObserver, connectionNew)
                JournalModel.sync(syncObserver, connectionNew)
                NewsModel.sync(syncObserver, connectionNew)

                syncObserver.finish()

                true
            }
        }
        catch (ex: Exception)
        {
            if(ex.message != null)
            {
                syncObserver.OnUpdated(ex.message!!)
            }

            false
        }
    }

    init
    {
        //Ip = Db["Settings.IP"]
    }

    fun reset()
    {
        InboxModel.clear()
        CalendarModel.clear()
        JournalModel.clear()
        FinanceModel.clear()
        RewardModel.clear()
        NewsModel.clear()
    }
}
